### ISSUE TITLE
Hash repeatability in cask XCode install

### ISSUE THREAD

**Issue by tmaone** [toonetown/homebrew-xcode#1]
Hi there, I was having some sdk issues compiling projects through xcode, so I decided to start everything from scratch via this tap, i.e. the non app-store xcode specifically, and then subsequently the legacy sdks. 

FIY here are the steps I've followed as indicated:

```
> open Xcode8.3.2.xip
> mkdir Xcode-tmp
> mv Xcode.app Xcode-tmp/Xcode.app
> hdiutil create -volname "Xcode"  -srcfolder Xcode-tmp   -ov -format UDZO Xcode8.3.2.dmg
>..................................................................................................................................
> created: xxxxxx/xcode-downloads/Xcode8.3.2.dmg
> rm -rf Xcode-tmp
> cask install xcode
> ==> Caveats 
> ==> Satisfying dependencies
> ==> Installing Formula dependencies from Homebrew
> toonetown/xcode/link-sdks ... ==> Cloning into 'xxxxxx/Library/Caches/Homebrew/link-sdks--git'...
> done
> complete
> ==> Downloading file:/xxxxxxxx/developer/xcode-downloads/Xcode8.3.2.dmg
> ############################################################## 100.0%
> ==> Verifying checksum for Cask xcode
> ==> Note: running "brew update" may fix sha256 checksum errors
> Error: sha256 mismatch
> Expected: dc804286b87acd1c9a4e3fb6916610bec0861be6b3bcaa3495a6a0297b32e7f6
> Actual: 49b0a99e556a918a92621a2b7da1e4f35bf267c9768eb82210ec2d67913ca972
> File: /xxxxxxx/Library/Caches/Homebrew/Cask/xcode--8.3.2.dmg
> To retry an incomplete download, remove the file above.
```

So, I guess there has to be some kind of way that we can all establish that the manually created dmg, somehow produces the same hash for everyone, where in fact, there can be several failing points (for mutual hash agreement) that can happen in the process. 

Any suggestions? Or you think due to the nature of the process that it might not be really worth investing time on it? On a another side, (and I haven't really looked into it) I'm aware that there exist hash database repositories for the OS installers, but none that I know of for the xcode app. 

Forgot to add the details of my product, to cross-verify in case I did something wrong:

```
> file Xcode8.3.2.dmg
> Xcode8.3.2.dmg: VAX COFF executable - version 5624
> ls -la@O Xcode8.3.2.dmg
> -rw-r--r--@ 1 user  staff  - 4949639266  2 Jun 23:44 Xcode8.3.2.dmg
>	com.apple.FinderInfo	        32
>  hdiutil verify Xcode8.3.2.dmg
> Checksumming Protective Master Boot Record (MBR : 0)…
> Protective Master Boot Record (MBR :: verified   CRC32 $80859BB7
> Checksumming GPT Header (Primary GPT Header : 1)…
>  GPT Header (Primary GPT Header : 1): verified   CRC32 $85B42E57
> Checksumming GPT Partition Data (Primary GPT Table : 2)…
> GPT Partition Data (Primary GPT Tabl: verified   CRC32 $BBCB003B
> Checksumming  (Apple_Free : 3)…
>                     (Apple_Free : 3): verified   CRC32 $00000000
> Checksumming EFI System Partition (C12A7328-F81F-11D2-BA4B-00A0C93EC93B : 4)…
> EFI System Partition (C12A7328-F81F-: verified   CRC32 $B54B659C
> Checksumming disk image (Apple_HFS : 5…..........................
>           disk image (Apple_HFS : 5): verified   CRC32 $99631250
> Checksumming  (Apple_Free : 6)…............................................................
>                     (Apple_Free : 6): verified   CRC32 $00000000
> Checksumming GPT Partition Data (Backup GPT Table : 7)…...........................................................
> GPT Partition Data (Backup GPT Table: verified   CRC32 $BBCB003B
> Checksumming GPT Header (Backup GPT Header : 8)…............................................................
>   GPT Header (Backup GPT Header : 8): verified   CRC32 $2B40DCD7............................................
> verified   CRC32 $9DAE390F
> hdiutil: verify: checksum of "Xcode8.3.2.dmg" is VALID
> hdiutil checksum -type SHA256 Xcode8.3.2.dmg
> ......................................................
> calculated SHA256 $1AAC9A79D153BE0C833BBA8058C4DE262B01041BED7C848CC7DD8BEDA15EE0F1
> sha256sum Xcode8.3.2.dmg
> 49b0a99e556a918a92621a2b7da1e4f35bf267c9768eb82210ec2d67913ca972  Xcode8.3.2.dmg
> shasum --algorithm 256 Xcode8.3.2.dmg **(What cask uses internally)** 
> 49b0a99e556a918a92621a2b7da1e4f35bf267c9768eb82210ec2d67913ca972  Xcode8.3.2.dmg
```

**Comment by toonetown** [comment]
Yeah - extracting the xip and creating the dmg will produce a different hash every time. I use this tap for my own purposes (syncing installations across a bunch of build servers), so you are running into an issue that I knew existed...but haven't had cause me pain yet. 

I have toyed with the idea of hosting the files publicly, but I don't think Apple would be too happy about me doing that - so they are privately hosted. 

It's possible to disable hash checking in the cask, I believe...but especially since you are doing compilation of code, in this case, you really need to make sure that signatures match.  

For example, I know that these signatures match "my" dmgs...because I built them. The problem that arises if I host them publicly is that "you" shouldn't trust them...because you shouldn't trust "me". 

I don't know a good way around this situation aside from forking this repo and maintaining your own hashes separately.  

Those hashes could possibly even all be in one file - which each forker could then just replace (and then the hashtable file would not be checked in to the repo).

**Comment by tmaone** [comment]
Yes, mutual agreement is the key. And evidently almost impossible to reach at this point. I was under the impression that they were intended for public use, sorry about that.  I mean the scripts and brew workflow are more than cool, and for me too, so. As it is right now, I'm not sure either a solution a-la hash db is worth the time of you first of all, and me as well, so I will fork and merge your upstream, or even PR, and if this issue becomes important, I'll see if I could think something about it to contribute back. Anyhow thanks for your prompt reply, feel free to close the issue, and I will be in touch when needed. Thanks again for your effort.

**Comment by tmaone** [comment]
Eventually, the workaround is to replace that hash in that cask as you correctly suggest, where after that everything rolls out smoothly:

```
==> Installing Formula dependencies from Homebrew
toonetown/xcode/link-sdks ... ==> Warning: toonetown/xcode/link-sdks 2 is already installed
done
complete
==> Downloading file:/xxxxxxx/developer/xcode-downloads/Xcode8.3.2.dmg
Already downloaded: /xxxxxxx/Library/Caches/Homebrew/Cask/xcode--8.3.2.dmg
==> Verifying checksum for Cask xcode
==> Installing Cask xcode
==> Moving App 'Xcode.app' to '/Applications/Xcode.app'.
==> Selecting default version of Xcode (may require sudo)
☉ With great power comes great responsibility
🔒
Error: Wrong password
🔒
==> Agreeing to license
==> Relinking SDKs
🍺  xcode was successfully installed!
```

Cheers

**Comment by toonetown** [comment]
Glad you got it working - and yes, that is the only workaround for now. 

Obviously, I should be more clear about the state of this tap on the readme. 😬

However, fee free to use them - I would love to get more exercise and use out of them...

Plus, if you would like to work out a pull request or something to manage the hashes externally, I would be more than willing to review and include it. The way I would envision it working is that you could provide a file (hashes.rb or something) that would just contain key-value pairs of each download binary. Each cask within this tap could then be able to include that file to load the hash for it to use. 

However, I am happy that things work with the workaround you used as well. Sorry it isn't more automatic.

**Comment by tmaone** [comment]
Nothing to be sorry about, honestly. You seem correct in the approach for handling the hashes, and to be precise, Im not entirely sure weather a per-file hash calculation or the .xip's hash should be more appropriate, if we were talking larger scale. For example, I have seen both examples in known repos, e.g. installer type hash (macOS-Security-and-Privacy-Guide/InstallESD_Hashes.csv) or a more individualised hash solution, way less prone to error, (for example using something like this andreasfink/dirhash or this JayBrown/Checksums) is the way to go. Of course the latter case if we consider the xip to be volatile. But let's see. 


Ok, so I think I got derailed by the brew hash, since in such taps here things are not going the conventional way the solution might be much simpler, for example take a look how the fastlane people go about verifying the authenticity of the downloaded xcode app here: 

```https://github.com/fastlane/fastlane/blob/master/fastlane/lib/fastlane/actions/verify_xcode.rb```

I think this something that can be looked into in a fairly shorter time. yay  
🎉

**Comment by toonetown** [comment]
The hash of the xip would be totally appropriate (because I "trust" Apple). However, I had a hard time scripting the unpacking of the xip archive - which is why I took the approach that I have. (FYI - if you install an "older" version of Xcode or an SDK that Apple provides as a dmg, it should work for you without needing to tweak the hash file)

Ideally, casks would "just work" on xip files directly (like they do with dmg, zip, tar.gz, pkg, etc). I don't know if the cask maintainers would consider supporting xip as a packaging format. If we could get xip support in homebrew-cask itself, it would be completely possible to just extract and install just like every other cask.

That would be the "best" solution, in my opinion (then, this tap would just have the hashes of the "official" Apple distributions)...but I haven't had the time to investigate what that would entail.
